packages/affiliates/libs/uniswap/models.ts

```diff
@@ -0,0 +1,169 @@
+import assert from "assert";
+import { exists, getPositionKey } from "./utils";
+import { ethers } from "ethers";
+
+type Id = string | number;
+type HasId = {
+  id?: Id;
+};
+interface MakeId<T> {
+  (obj: T): string;
+}
+
+// Generic table with common functions, store is a map now, but we could replace this with any kind of
+// data store with some minor modifications, as long as it has, set, get, delete, has and values
+export function Table<T extends HasId>(config: { makeId: MakeId<T>; type: string }, store: Map<Id, T>) {
+  const { makeId, type } = config;
+  async function create(data: T) {
+    const id = makeId(data);
+    assert(!(await has(id)), `${type} exists`);
+    return set({ id, ...data });
+  }
+  async function set(data: T) {
+    assert(data.id, `${type} requires id`);
+    await store.set(data.id, { ...data });
+    return data;
+  }
+  async function get(id: string) {
+    assert(await store.has(id), `${type} does not exist`);
+    return store.get(id) as T;
+  }
+  async function has(id: string) {
+    return store.has(id);
+  }
+  async function list() {
+    return [...store.values()];
+  }
+  async function forEach(cb: (value: T, index: number, array: T[]) => void) {
+    const all = await list();
+    all.forEach(cb);
+  }
+  async function update(id: string, data: any = {}) {
+    const got = await get(id);
+    return set({ ...got, ...data });
+  }
+  return {
+    create,
+    set,
+    get,
+    has,
+    update,
+    forEach,
+    list,
+    store
+  };
+}
+
+export type NftPosition = {
+  id?: string;
+  tokenId: string;
+  amount?: string;
+  amount0?: string;
+  amount1?: string;
+  token0?: string;
+  token1?: string;
+  liquidity?: string;
+  feeGrowthInside0LastX128?: string;
+  feeGrowthInside1LastX128?: string;
+  tokensOwed0?: string;
+  tokensOwed1?: string;
+  pool?: string;
+};
+
+// Nft positions are gathered from the nft contract, and require slightly different indexing than the positions in pool contract
+export const NftPositions = () => {
+  function makeId(data: NftPosition) {
+    return data.tokenId;
+  }
+  const store = new Map<Id, NftPosition>();
```

Comment:
nit: why construct this outside? Why not construct this within the Table constructor since it's the same every time?

Reply:
id like to inject a different type of store in the table eventually.  Would eventually need a store interface independent of the JS Map that all stores could conform to, so right now id like to leave that part out of the table. Essentially the table can become agnostic to the store technology, and you would be able to plug in different store adapters.